src/hammer-vlsi/hammer_utils/spice_utils.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+#  spice_utils.py
+#  Misc Verilog utilities
+#
+#  See LICENSE for licence details.
+
+import re
+from typing import Tuple, Optional, List, Set, Dict
+from enum import Enum
+
+class SpiceUtils:
+
+    # Multilines in spice start with a + character
+    multiline_pattern = re.compile("(?:\s*\n\+)+\s*", flags=re.DOTALL)
+    module_def_pattern = re.compile("^(\.subckt\s+)(?P<dname>[^\s]+)(\s+.*)$", flags=re.IGNORECASE|re.MULTILINE)
+    module_inst_pattern = re.compile("^(x.*?\s)(?P<mname>[^\s]+)(\s*)$", flags=re.IGNORECASE|re.MULTILINE)
+    end_module_pattern = re.compile("^\.ends.*$", flags=re.IGNORECASE|re.MULTILINE)
+    tokens = [
+        ('SUBCKT', module_def_pattern.pattern),
+        ('ENDS', end_module_pattern.pattern),
+        ('INST', module_inst_pattern.pattern)
+    ]
+    token_regex = re.compile('|'.join('(?P<%s>%s)' % t for t in tokens), flags=re.IGNORECASE|re.MULTILINE)
+
+    @staticmethod
+    def uniquify_spice(sources: List[str]) -> List[str]:
+        """
+        Uniquify the provided SPICE sources. If a module name exists in multiple files, each duplicate module will be
+        renamed. If the original name of a duplicated module is used in a file where it is not defined, an
+        exception is thrown as it is impossible to know which renamed module to use. For example:
+
+        - file1.sp defines B and A, which instantiates B
+        - file2.sp defines B and C, which instantiates B
+
+        B would be renamed to B_0 in file1.sp
+        B would be renamed to B_1 in file2.sp
+
+        However, if file3.sp defines D, which instantiates B, an exception would be raised because we don't know if we
+        should chose B_0 or B_1.
+
+        :param sources: A list of the contents of SPICE source files (not the filenames)
+        :return: A list of SPICE sources with unique modules
+        """
+        sources_no_multiline = [SpiceUtils.remove_multilines(s) for s in sources]
```

Comment:
Bike-shedding: if there is no lambda being created it's slightly more efficient to use `map`

```suggestion
        sources_no_multiline = list(map(SpiceUtils.remove_multilines, sources))
```

Reply:
These all take a long time, so I'm fine with the extra few microseconds. The improved readability of the list comprehension is worth it.